### DISCUSSION TITLE
cub3d

### DISCUSSION THREAD

**Discussion by SandraKanna** [denise-sandra/minishell#30] [General]
https://prod.liveshare.vsengsaas.visualstudio.com/join?DC411FF85EEBF2BD420569ABC1EDC6E8FAE1

**Comment by SandraKanna** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?A7C6BEACE9BD2CCE6A28A85DC5BFEF6BB7EA

**Reply by SandraKanna** [reply]
https://prod.liveshare.vsengsaas.visualstudio.com/join?A7C6BEACE9BD2CCE6A28A85DC5BFEF6BB7EA

**Comment by SandraKanna** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?2A918F8449084F04B5B9401F5CA37D040B90

**Comment by SandraKanna** [comment]
#ifndef CUB3D_H
# define CUB3D_H

# include <stdlib.h>
# include <fcntl.h>
# include <stdio.h>
# include <unistd.h>
# include <math.h>
# include "libft_extended.h"
# include "MLX42/MLX42.h"
# include "MLX42/MLX42_Int.h"

# define WINDOW_WIDTH 960
# define WINDOW_HEIGHT 540
// # define TILE_SIZE 60

typedef struct	s_map
{
	char 			*value;
	struct s_map	*next;
}	t_map;

typedef struct s_context
{
	t_map	*map;
	void	*no;
	void	*so;
	void	*we;
	void	*ea;
	int		f_color[3];
	int		c_color[3];
	// int		p_coord[2];
}	t_context;

typedef struct s_player
{
	double	x;
	double	y;
	double	dir_x;
	double	dir_y;
	double	plane_x;
	double	plane_y;

}	t_player;

typedef struct s_game
{
	t_context		*context;
	t_player		player;
	//mlx_texture_t	*texture;
	void			*mlx42;
}	t_game;

void	ft_err(char *msg, t_game *game);
void	get_context(t_context *map, char *map_file);
void	clean_game(t_game *game);
void	free_maplst(t_map **lst);

#endif

**Comment by SandraKanna** [comment]
if (line[i] == 32)
				i++;
			if (line[i] == 'S' && line[i + 1] == 'O')
			{
				i = i + 2;
				while (line[i] && line[i] == 32)
					i++;
				context->so = malloc((ft_strlen(line) - i + 1) * sizeof(void));
				if (!context->so)
					context->so = line + i;
			}
			else if (line[i] == 'N' && line[i + 1] == 'O')
			{
				i = i + 2;
				while (line[i] && line[i] == 32)
					i++;
				context->no = line + i;
			}
			else if (line[i] == 'W' && line[i + 1] == 'E')
			{
				i = i + 2;
				while (line[i] && line[i] == 32)
					i++;
				context->we = line + i;
			}
			else if (line[i] == 'E' && line[i + 1] == 'A')
			{
				i = i + 2;
				while (line[i] && line[i] == 32)
					i++;
				context->ea = line + i;
			}
			else if (line[i] == 'F' && line[i + 1])
			{
				i = i + 2;
				while (line[i] && line[i] == 32)
					i++;
				tmp = line + i;
				rgb = ft_split(tmp, ',');
				context->f_color[0] = ft_atoi(rgb[0]);
				context->f_color[1] = ft_atoi(rgb[1]);
				context->f_color[2] = ft_atoi(rgb[2]);
			}
			else if (line[i] == 'C' && line[i + 1])
			{
				i = i + 2;
				while (line[i] && line[i] == 32)
					i++;
				tmp = line + i;
				rgb = ft_split(tmp, ',');
				context->c_color[0] = ft_atoi(rgb[0]);
				context->c_color[1] = ft_atoi(rgb[1]);
				context->c_color[2] = ft_atoi(rgb[2]);
			}

**Comment by derjavec** [comment]
static int     read_nodes(char *line)
{
	int  i;

	i = 0;
	while ((line[i] == 32 || line[i] == 11 \
		|| line[i] == 12 || line[i] == 13))
		i++;
	if (ft_strncmp("NO", line + i, 2) == 0)
		return (1);
	if (ft_strncmp("SO", line + i, 2) == 0)
		return (1);
	if (ft_strncmp("WE", line + i, 2) == 0)
		return (1);
	if (ft_strncmp("EA", line + i, 2) == 0)
		return (1);
	if (ft_strncmp("\n", line + i, 1) == 0)
		return (1);
	return (0);
}

void check_map(t_game *game)
{
	t_map *map;
	t_map *prev = NULL;
	t_map *next;

	map = game->context->map;
	while (map)
	{
		next = map->next;
		printf("value: %s\n", map->value);
		if (read_nodes(map->value) == 1)
		{
			// ft_mapdelone(map);
			
			free (map->value);
			free (map);
			if (prev)
				map = prev->next;
			else
			
			map->next = next;
		}
		prev = map;
		map = map->next;
	}
	// t_map *print = game->context->map;
	// 	while(print)
	// 	{
	// 		printf("maplist: %s\n", print->value);
	// 		print = print->next;
	// 	}
}

**Reply by SandraKanna** [reply]
void check_map(t_game *game)
{
    t_map *map;
    t_map *prev = NULL;
    t_map *next;

    map = game->context->map;
    while (map)
    {
        next = map->next; // Guarda el siguiente nodo
        if (read_nodes(map->value) == 1) // Si se cumple la condición
        {
            if (prev) // Si no es el primer nodo
                prev->next = next;
            else // Si es el primer nodo, actualiza la cabeza de la lista
                game->context->map = next;

            free(map->value); // Libera el contenido del nodo
            free(map); // Libera el nodo

            // No se avanza prev porque el nodo actual fue eliminado
            map = next; // Continuar con el siguiente nodo
        }
        else
        {
            prev = map; // Avanza prev si no se eliminó el nodo actual
            map = next; // Avanza al siguiente nodo
        }
    }
}

**Comment by derjavec** [comment]
https://github.com/erivero-p/cub3d-Raycast-Tutorial

**Comment by SandraKanna** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?C30F7E66A7B702D45262C13DB9D01D523E1C

**Comment by SandraKanna** [comment]
11111111111111111111111 1
111100000011  100000000000011
    1011000011110000000000001
1111000000000000000000001111111
111111111011000001110000000000011
100000000011000001110111111111111
111101111111111011100000010001
1111011111111 11011101010000101
110000001101010111000000100011
11110000000000001100000010001
    10000000000000001101010011
11110001110101011111N1111000111
110000111 1110101 101111010001
11111111111111111111111111111

**Comment by SandraKanna** [comment]
1111111111111111111111111
        1000000000110000000000001
        1011000001110000000000001
        1001000000000000000000001
1111111110110000011100000001000001
100000000011000001110111111111111
111101111111110111000000100N1
1111011111111101110101001001
11000000110101011100000010001
10000000000000001100000010001
10000000000000001111010010001
110000011101010111110111100111
11110111 1110101 101111010001
11111111 1111111 111111111111

**Comment by SandraKanna** [comment]
static int	can_move(t_game *game, double new_x, double new_y)
{
	int	dir_x;
	int	dir_y;

	if (new_x < game->player.old_x)
		dir_x = -1;
	else
		dir_x = 1;
	if (new_y < game->player.old_y)
		dir_y = -1;
	else
		dir_y = 1;
	printf(" %d %d %c\n", (int)floor(new_y + (D_ROUND * dir_y)),(int)floor(new_x + (D_ROUND * dir_x)), game->map[(int)floor(new_y + (D_ROUND * dir_y))][(int)floor(new_x + (D_ROUND * dir_x))]);
	if (game->map[(int)floor(new_y + (D_ROUND * dir_y))][(int)floor(new_x + (D_ROUND * dir_x))] != '1')
		return (1);
	return (0);
}

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?103E9BF3FBAD2F5724D931243D57F07DC6A3

**Comment by SandraKanna** [comment]
nuevo archivo: parse_bonus.c


#include "cub3d_bonus.h"

static void	char_duplicate(char c, int *door, int *pope, t_game *game)
{
	if (c == '2')
		(*door)++;
	if (c == 'Z')
		(*pope)++;
	if (*door > 1)
		ft_err("Invalid bonus map: duplicated door\n", game);
	if (*pope > 1)
		ft_err("Invalid bonus map: duplicated pope\n", game);
}

static void	check_valid_chars_bonus(t_context **cur, t_game *game)
{
	int	i;
	int	door;
	int	pope ;

	i = 0;
	door = 0;
	pope = 0;
	while ((*cur)->value[i])
	{
		char_duplicate((*cur)->value[i], &door, &pope, game);
		if ((*cur)->value[i] != ' ' && (*cur)->value[i] != '1' && \
		(*cur)->value[i] != '2' && (*cur)->value[i] != 'N' && \
		(*cur)->value[i] != 'S' && (*cur)->value[i] != 'E' && \
		(*cur)->value[i] != 'W' && (*cur)->value[i] != '0' && \
		(*cur)->value[i] != 'Z' && (*cur)->value[i] != '\n')
			ft_err("Invalid character in map\n", game);
		if ((*cur)->value[i] == ' ')
			(*cur)->value[i] = 'X';
		if ((*cur)->value[i] == '\n')
			(*cur)->value[i] = '\0';
		i++;
	}
}

void	parse_map_bonus(t_game *game)
{
	t_context	*cur;
	int			rows;
	int			dir;

	cur = game->list;
	while (cur)
	{
		if (ft_strncmp("\n", cur->value, ft_strlen(cur->value)) == 0)
			ft_err("Invalid map: empty line\n", game);
		check_valid_chars_bonus(&cur, game);
		cur = cur->next;
	}
	game->map = list_to_tab(game);
	if (!game->map)
		ft_err("Failed to malloc\n", game);
	rows = 0;
	while (game->map[rows])
		rows++;
	dir = get_player_pos(game, game->map, rows);
	floodfill(game, game->player.x, game->player.y, rows);
	free_tab(game->map);
	game->map = list_to_tab(game);
	initialize_player_dir(game, dir);
	clean_context_lst(&game->list);
	game->list = NULL;
}

**Comment by SandraKanna** [comment]
mlx_image_t	*display_texture(mlx_t *mlx, mlx_texture_t *texture, int x, int y)
{
	mlx_image_t	*img;

	img = mlx_new_image(mlx, texture->width, texture->height);
	if (!img)
		return (NULL);
	ft_memcpy(img->pixels, texture->pixels, \
		texture->width * texture->height * 4);
	mlx_image_to_window(mlx, img, x, y);
	return (img);
}

**Comment by Domingos-Franco** [comment]
Somebody can recommend some materials, to understand very well this project and concepts of graphics computation?
i have been implement algorithm  ray_casting to get the projection, pov and movements... but i not understand how this process really works.
